### ISSUE TITLE
Syncing from multiple BTCD backends can cause slow sync with out of order cfheaders being stashed

### ISSUE THREAD

**Issue by mrfelton** [lightningnetwork/lnd#1752]
### Background

Using neutrino, it is possible to connect lnd to multiple BTCD backends by setting the `--neutrino.connect` option more than once. When performing a fresh initial sync this can cause  cfheaders to be received out of order which has a knock on effect of making the synchronisation process considerably slower than if it were only syncing from a single host.

Here are some experts from the logs that highlight the issue:

```
[INF] BTCN: Starting cfheaders sync at block_height=1384261, hash=000000000001e2d21cabb2927085ed4418db8fc63fa1d826cefab60c3c0fee5c
[INF] BTCN: Starting cfheaders sync for filter_type=0
[DBG] BTCN: Sending getcfcheckpt to 35.231.86.179:18333 (outbound)
[DBG] BTCN: Sending getcfcheckpt to 35.196.85.107:18333 (outbound)
[INF] LNWL: Done catching up block hashes
[INF] LNWL: Started rescan from block 000000000001e2d21cabb2927085ed4418db8fc63fa1d826cefab60c3c0fee5c (height 1384261) for 1 address
[DBG] BTCN: Received cfcheckpt from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received unhandled message of type cfcheckpt from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received cfcheckpt from 35.231.86.179:18333 (outbound)
[DBG] BTCN: Received unhandled message of type cfcheckpt from 35.231.86.179:18333 (outbound)
[INF] BTCN: Fetching set of checkpointed cfheaders filters from height=0, hash=21584579b7eb08997773e5aeff3a7f932700042d0ed2a6129012b7d7ae81b750
[INF] BTCN: Starting to query for cfheaders from checkpoint_interval=0
[INF] BTCN: Attempting to query for 1384 cfheader batches
[DBG] BTCN: Sending getcfheaders to 35.231.86.179:18333 (outbound)
[DBG] BTCN: Sending getcfheaders to 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.231.86.179:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=1 to height=1001
[DBG] BTCN: Writing filter headers up to height=1000
[INF] BTCN: Verified 1 filter header in the last 3m53.05s (height 1, 2011-02-02 23:22:08 +0000 UTC)
[DBG] BTCN: Sending getcfheaders to 35.231.86.179:18333 (outbound)
[DBG] BTCN: Waiting for filter headers (height=1000) to catch up the rescan start (height=1384261)
[INF] LNWL: Catching up block hashes to height 1384261, this might take a while
[INF] LNWL: Done catching up block hashes
[INF] LNWL: Finished rescan for 1 address (synced to block 000000000001e2d21cabb2927085ed4418db8fc63fa1d826cefab60c3c0fee5c, height 1384261)
[DBG] BTCN: Received cfheaders from 35.231.86.179:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=2001 to height=3001
[DBG] BTCN: Got response for headers at height=2001, only at height=1000, stashing
[DBG] BTCN: Sending getcfheaders to 35.231.86.179:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.231.86.179:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=1001 to height=2001
[DBG] BTCN: Writing filter headers up to height=2000
[DBG] BTCN: Sending getcfheaders to 35.196.85.107:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=3001 to height=4001
[DBG] BTCN: Got response for headers at height=3001, only at height=2000, stashing
[DBG] BTCN: Sending getcfheaders to 35.231.86.179:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.231.86.179:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=5001 to height=6001
[DBG] BTCN: Got response for headers at height=5001, only at height=2000, stashing
[DBG] BTCN: Sending getcfheaders to 35.231.86.179:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=4001 to height=5001
[DBG] BTCN: Got response for headers at height=4001, only at height=2000, stashing
[DBG] BTCN: Sending getcfheaders to 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.231.86.179:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=6001 to height=7001
[DBG] BTCN: Got response for headers at height=6001, only at height=2000, stashing
[DBG] BTCN: Sending getcfheaders to 35.231.86.179:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=7001 to height=8001
[DBG] BTCN: Got response for headers at height=7001, only at height=2000, stashing
```

Here is a subset of the above lines for clarity:

```
[DBG] BTCN: Got cfheaders from height=1 to height=1001
[DBG] BTCN: Writing filter headers up to height=1000
[DBG] BTCN: Got cfheaders from height=2001 to height=3001
[DBG] BTCN: Got response for headers at height=2001, only at height=1000, stashing
[DBG] BTCN: Got cfheaders from height=1001 to height=2001
[DBG] BTCN: Writing filter headers up to height=2000
[DBG] BTCN: Got cfheaders from height=3001 to height=4001
[DBG] BTCN: Got response for headers at height=3001, only at height=2000, stashing
[DBG] BTCN: Got cfheaders from height=5001 to height=6001
[DBG] BTCN: Got response for headers at height=5001, only at height=2000, stashing
[DBG] BTCN: Got cfheaders from height=4001 to height=5001
[DBG] BTCN: Got response for headers at height=4001, only at height=2000, stashing
[DBG] BTCN: Got cfheaders from height=6001 to height=7001
[DBG] BTCN: Got response for headers at height=6001, only at height=2000, stashing
[DBG] BTCN: Got cfheaders from height=7001 to height=8001
[DBG] BTCN: Got response for headers at height=7001, only at height=2000, stashing
```

Notice that they are out of order.

As soon as it has received some of the headers out of order as above, it starts "stashing" the data as it receives it. In the above example it has only written the filters up to block 2000 and from there on they are just being stashed. This will continue until every filter header has been fetched and only then will it go back to block 2000 and start "writing" the filter headers.

On testnet, this adds 2-3 minutes extra on the sync process as the fetching and writing of the filter headers happens sequentially instead or in parallel. Effectively it has to do a second pass over all of the filter headers. First fetch and stash them all, then write them all.

For contrast, if you only sync from a single host then this behaviour is not observed and each set of headers is written immediately after it has been fetched (there is no stashing), resulting in a faster sync time.

### Your environment

* version of `lnd`: lnd version 0.4.2-beta commit=147596047f5fac79cba4128a6a2aa47bce2e6d7c
* which operating system (`uname -a` on *Nix): Darwin
* version of `btcd`, `bitcoind`, or other backend: BTCD master

### Steps to reproduce

Set ` --neutrino.connect` twice (two different btcd nodes) and initialise a fresh wallet and sync.

```
lnd ...  --neutrino.connect=btcdnode1 --neutrino.connect=btcdnode2 ...
```

### Expected behaviour

The time to sync from multiple hosts should be the same or faster than from a single host. Filter headers should be written as they are fetched. Here are the log entries from a sync from a single node where you can see everything being done in the correct order and no "stashing". Each set of headers is written immediately after it has been received, which is about 50% faster.

```
[INF] BTCN: Starting cfheaders sync at block_height=1384269, hash=000000000000003e8b5add485d78037acfc29fed4bdf1ddb1ff92cee661bdb17
[INF] BTCN: Starting cfheaders sync for filter_type=0
[DBG] BTCN: Sending getcfcheckpt to 35.196.85.107:18333 (outbound)
[INF] LNWL: Done catching up block hashes
[INF] LNWL: Started rescan from block 000000000000003e8b5add485d78037acfc29fed4bdf1ddb1ff92cee661bdb17 (height 1384269) for 1 address
[DBG] BTCN: Received cfcheckpt from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received unhandled message of type cfcheckpt from 35.196.85.107:18333 (outbound)
[INF] BTCN: Fetching set of checkpointed cfheaders filters from height=0, hash=21584579b7eb08997773e5aeff3a7f932700042d0ed2a6129012b7d7ae81b750
[INF] BTCN: Starting to query for cfheaders from checkpoint_interval=0
[INF] BTCN: Attempting to query for 1384 cfheader batches
[DBG] BTCN: Sending getcfheaders to 35.196.85.107:18333 (outbound)
[DBG] BTCN: Sending ping to 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received ping from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Sending pong to 35.196.85.107:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=1 to height=1001
[DBG] BTCN: Writing filter headers up to height=1000
[INF] BTCN: Verified 1 filter header in the last 4m0.41s (height 1, 2011-02-02 23:22:08 +0000 UTC)
[DBG] BTCN: Sending getcfheaders to 35.196.85.107:18333 (outbound)
[INF] LNWL: Catching up block hashes to height 1384269, this might take a while
[DBG] BTCN: Waiting for filter headers (height=1000) to catch up the rescan start (height=1384269)
[INF] LNWL: Done catching up block hashes
[INF] LNWL: Finished rescan for 1 address (synced to block 000000000000003e8b5add485d78037acfc29fed4bdf1ddb1ff92cee661bdb17, height 1384269)
[DBG] BTCN: Received pong from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=1001 to height=2001
[DBG] BTCN: Writing filter headers up to height=2000
[DBG] BTCN: Sending getcfheaders to 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=2001 to height=3001
[DBG] BTCN: Writing filter headers up to height=3000
[DBG] BTCN: Sending getcfheaders to 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=3001 to height=4001
[DBG] BTCN: Writing filter headers up to height=4000
[DBG] BTCN: Sending getcfheaders to 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=4001 to height=5001
[DBG] BTCN: Writing filter headers up to height=5000
[DBG] BTCN: Sending getcfheaders to 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=5001 to height=6001
[DBG] BTCN: Writing filter headers up to height=6000
[DBG] BTCN: Sending getcfheaders to 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=6001 to height=7001
[DBG] BTCN: Writing filter headers up to height=7000
[DBG] BTCN: Sending getcfheaders to 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=7001 to height=8001
[DBG] BTCN: Writing filter headers up to height=8000
```

### Actual behaviour

Sync time is slower from multiple hosts. Filter headers are stashed as they are fetched and then written only after they have all been fetched.

**Comment by Roasbeef** [comment]
I'm unable to repro this. Connecting to multiple nodes speeds up sync as we're able to distribute the queries amongst the range of peers, so we can actually parallelize download. I'd check the logs (on a higher setting) to see if that request is actually being serviced by the peer. 

We've made a ton of optimizations to the cfheader scheduler, but as this isn't really our primary focus atm, we put in on hold in order to pursue higher priority items.

**Comment by Roasbeef** [comment]
Considering migrating this issue to the `neutrino` repo where it really belongs.

**Comment by mrfelton** [comment]
Hey @Roasbeef . I just updated to the most recent BTCD and LND which includes a more recent version of neutrino and can confirm that this issue does still exist.

Now the behaviour is slightly different from before, in that when syncing from multiple BTCD hosts the sync time for `getcfheaders` is notably faster. I can see from the logs that it is fetching different batches from each host in parallel.

However, it does still have this issue where it receives some of the filter headers out of order and then starts to stash everything until all of the headers have been fetched at which point it goes back to the start and process everything again.

In my testing it is in the order of 30% slower to sync from 2 BTCD hosts. Here are the sync times

**1 host:**
```
getheaders: 4:00
getcfheaders: 6:14
```

**2 hosts:**
```
getheaders: 4:00
getcfheaders: 3:21
getcfheaders (round 2 - writing filter headers): 5:50
```

And an except from the logs where you can see it receiving the first two batches of data out of order and subsequently stashing everything until a second pass.

```
[INF] BTCN: Starting cfheaders sync for filter_type=0
[DBG] BTCN: Sending getcfcheckpt to 35.196.85.107:18333 (outbound)
[DBG] BTCN: Sending getcfcheckpt to 35.231.86.179:18333 (outbound)
[INF] LNWL: Done catching up block hashes
[INF] LNWL: Started rescan from block 000000000000007c83a5881241f5c609b3e615ec39de803d87e5b061fd894c3b (height 1409811) for 1 address
[DBG] BTCN: Received cfcheckpt from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received unhandled message of type cfcheckpt from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received cfcheckpt from 35.231.86.179:18333 (outbound)
[DBG] BTCN: Received unhandled message of type cfcheckpt from 35.231.86.179:18333 (outbound)
[INF] BTCN: Fetching set of checkpointed cfheaders filters from height=0, hash=21584579b7eb08997773e5aeff3a7f932700042d0ed2a6129012b7d7ae81b750
[INF] BTCN: Starting to query for cfheaders from checkpoint_interval=0
[INF] BTCN: Attempting to query for 1409 cfheader batches
[DBG] BTCN: Sending getcfheaders to 35.231.86.179:18333 (outbound)
[DBG] BTCN: Sending getcfheaders to 35.196.85.107:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.231.86.179:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=1001 to height=2001
[DBG] BTCN: Got response for headers at height=1001, only at height=0, stashing
[DBG] BTCN: Sending getcfheaders to 35.231.86.179:18333 (outbound)
[DBG] BTCN: Received cfheaders from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=1 to height=1001
[DBG] BTCN: Writing filter headers up to height=1000
[INF] BTCN: Verified 1 filter header in the last 4m12.91s (height 1, 2011-02-02 23:22:08
[DBG] BTCN: Sending getcfheaders to 35.196.85.107:18333 (outbound)
[INF] LNWL: Catching up block hashes to height 1409811, this might take a while
[DBG] BTCN: Waiting for filter headers (height=1000) to catch up the rescan start (height=1409811)
[INF] LNWL: Done catching up block hashes
[INF] LNWL: Finished rescan for 1 address (synced to block 000000000000007c83a5881241f5c609b3e615ec39de803d87e5b061fd894c3b, height 1409811)
[DBG] BTCN: Received cfheaders from 35.196.85.107:18333 (outbound)
[DBG] BTCN: Got cfheaders from height=3001 to height=4001
[DBG] BTCN: Got response for headers at height=3001, only at height=1000, stashing
```

In this example, it got Got cfheaders from height=1001 to height=2001 before it got height=1 to height=1001

Happy to continue this discussion over in the neutrino issues if you want to move it there.

**Comment by mrfelton** [comment]
For context, the both of the BTCD hosts addresses in question were behind a load balancer, with ClientIP based Session Affinity (2 IP addresses, each fronting 2 BTCD nodes). In order to rule out the load balancer as an issue I have reduced each replica set to a size of one - so there is strictly one BTCD node at each IP address. The issue though is still reproducible every time.

**Comment by Roasbeef** [comment]
Closing to instead track on the neutrino repo, thanks!
